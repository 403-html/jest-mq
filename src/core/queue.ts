type Message<T = any> = { type: string | undefined; id: number } & T;
type MessageHandler<T = any> = (message: Message<T>) => Promise<void> | void;

export class MessageQueue<T = any> {
  private sentMessages: Message<T>[] = [];
  private receivedMessages: Message<T>[] = [];
  private handlers: Map<string | undefined, MessageHandler<T>[]> = new Map();
  private messageCount: number = 0;

  constructor(public name: string) {
    if (!name) {
      throw new Error("Queue name is required");
    }
    if (name.length > 255) {
      throw new Error("Queue name is too long");
    }
    if (name.includes(" ")) {
      throw new Error("Queue name cannot contain spaces");
    }
  }

  getQueue(): {
    name: string;
    sentMessages: Message<T>[];
    receivedMessages: Message<T>[];
    handlers: Map<string | undefined, MessageHandler<T>[]>;
  } {
    return {
      name: this.name,
      sentMessages: [...this.sentMessages],
      receivedMessages: [...this.receivedMessages],
      handlers: new Map(this.handlers),
    };
  }

  clear(): void {
    this.sentMessages = [];
    this.receivedMessages = [];
    this.handlers.clear();
    this.messageCount = 0;
  }

  private async processHandlers(messageWithId: Message<T>): Promise<void> {
    const handlers = [
      ...(this.handlers.get(messageWithId.type) || []),
      ...(this.handlers.get(undefined) || []),
    ];

    await Promise.all(handlers.map((handler) => handler(messageWithId)));
  }

  sendMessage(message: T): number {
    const messageWithId: Message<T> = {
      ...message,
      type: (message as any).type,
      id: this.messageCount++,
    };
    this.sentMessages.push(messageWithId);
    this.processHandlers(messageWithId);
    return messageWithId.id;
  }

  receiveMessage(messageType?: string, autoAck = true): Message<T> | undefined {
    const messageIndex = messageType
      ? this.sentMessages.findIndex((m) => m.type === messageType)
      : 0;
    if (messageIndex === -1) {
      return undefined;
    }

    const message = this.sentMessages[messageIndex];

    if (autoAck) {
      this.sentMessages.splice(messageIndex, 1);
      this.receivedMessages.push(message);
    }

    return message;
  }

  onMessage(
    messageType: string | undefined,
    handler: MessageHandler<T>,
  ): () => void {
    if (!this.handlers.has(messageType)) {
      this.handlers.set(messageType, []);
    }
    this.handlers.get(messageType)?.push(handler);

    return () => this.offMessage(messageType, handler);
  }

  offMessage(
    messageType: string | undefined,
    handler: MessageHandler<T>,
  ): void {
    if (this.handlers.has(messageType)) {
      const handlersForType = this.handlers.get(messageType)!;
      this.handlers.set(
        messageType,
        handlersForType.filter((h) => h !== handler),
      );
    }
  }
}
