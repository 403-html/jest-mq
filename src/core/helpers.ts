import { Message, MessageQueue } from "./queue";

/**
 * Waits asynchronously for a specific type of message from the given message queue within a specified timeout period.
 *
 * @template T - The type of the message.
 * @param {Object} params - The parameters for the function.
 * @param {MessageQueue<T>} params.queue - The message queue to receive messages from.
 * @param {string} params.messageType - The type of message to wait for.
 * @param {number} [params.timeout=3000] - The maximum time to wait for the message, in milliseconds. Defaults to 3000 ms.
 * @returns {Promise<Message<T>>} A promise that resolves with the received message if it arrives within the timeout period, or rejects with an error if the timeout is reached.
 * @throws {Error} If the timeout is reached before a message of the specified type is received.
 */
export async function expectMessage<T = any>({
  queue,
  messageType,
  timeout = 3000,
}: {
  queue: MessageQueue<T>;
  messageType: string;
  timeout?: number;
}): Promise<Message<T>> {
  return new Promise((resolve, reject) => {
    const startTime = Date.now();
    const intervalId = setInterval(() => {
      const message = queue.receiveMessage(messageType);
      if (message) {
        clearInterval(intervalId);
        resolve(message);
      } else if (Date.now() - startTime > timeout) {
        clearInterval(intervalId);
        reject(new Error(`Timeout waiting for message type: ${messageType}`));
      }
    }, 10);
  });
}
