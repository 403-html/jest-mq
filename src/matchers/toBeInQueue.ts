import { MessageQueue } from "../core/queue";
import { matcherHint, printReceived, printExpected } from "jest-matcher-utils";

export const toBeInQueue = function (
  this: jest.MatcherContext,
  received: MessageQueue<any>,
  expectedMessage: any,
) {
  const queue = received.getQueue();
  const messageIsInQueue = queue.sentMessages.some((message) => {
    return this.equals(
      // Exclude id from comparison, it's provided by the queue
      { ...message, id: undefined },
      expectedMessage,
    );
  });

  return {
    pass: messageIsInQueue,
    message: () => {
      const hint = matcherHint(".toBeInQueue", "received", "expectedMessage");
      const receivedStr = printReceived(expectedMessage);
      const expectedStr = printExpected(queue.sentMessages);
      return `${hint}
      
      Expected message to be in queue:
        ${receivedStr}
      Received:
        ${expectedStr}`;
    },
  };
};
