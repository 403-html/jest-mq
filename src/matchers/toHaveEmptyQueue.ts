import { MessageQueue } from "../core/queue";
import { matcherHint, printReceived, printExpected } from "jest-matcher-utils";

export const toHaveEmptyQueue = function (
  this: jest.MatcherContext,
  received: MessageQueue<any>,
) {
  const queue = received.getQueue();
  const queueIsEmpty = queue.sentMessages.length === 0;

  return {
    pass: queueIsEmpty,
    message: () => {
      const hint = matcherHint(".toHaveEmptyQueue", "received", "");
      const receivedStr = printReceived(queue.sentMessages);
      return `${hint}
      
      Expected queue to be empty, but it contains:
        ${receivedStr}`;
    },
  };
};
