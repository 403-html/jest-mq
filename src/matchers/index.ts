import { toBeInQueue } from "./toBeInQueue";

declare global {
  namespace jest {
    interface Matchers<R> {
      toBeInQueue(expectedMessage: any): R;
    }
  }
}

expect.extend({
  toBeInQueue,
});
