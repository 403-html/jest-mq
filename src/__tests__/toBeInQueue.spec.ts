import "../matchers";
import { MessageQueue } from "../core/queue";

describe("toBeInQueue", () => {
  let queue: MessageQueue;

  beforeEach(() => {
    queue = new MessageQueue("test");
  });

  afterEach(() => {
    queue.clear();
  });

  it("should pass if message is in queue", () => {
    const message = { type: "test", payload: "test" };
    queue.sendMessage(message);
    expect(queue).toBeInQueue(message);
  });

  it("should fail if message is not in queue", () => {
    const message = { type: "test", payload: "test" };
    expect(() =>
      expect(queue).toBeInQueue(message),
    ).toThrowErrorMatchingSnapshot();
  });
});
