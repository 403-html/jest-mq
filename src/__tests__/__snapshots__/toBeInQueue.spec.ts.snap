// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toBeInQueue should fail if message is not in queue 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeInQueue([22m[32mexpectedMessage[39m[2m)[22m
      
      Expected message to be in queue:
        [31m{\\"payload\\": \\"test\\", \\"type\\": \\"test\\"}[39m
      Received:
        [32m[][39m"
`;
